src/vdaf/hits.rs

```diff
@@ -0,0 +1,741 @@
+    }
+
+    // Generate IDPF shares of the data and authentication vectors.
+    let mut data = IntoIter::new(I::gen(input, std::iter::repeat(I::Field::one()))?);
+    let mut auth = IntoIter::new(I::gen(input, auth_rand)?);
+
+    Ok([
+        InputShareMessage {
+            data: data.next().unwrap(),
+            auth: auth.next().unwrap(),
+            sketch_start_seed: leader_sketch_start_seed,
+            sketch_next: Share::Leader(leader_sketch_next),
+        },
+        InputShareMessage {
+            data: data.next().unwrap(),
+            auth: auth.next().unwrap(),
+            sketch_start_seed: helper_sketch_start_seed,
+            sketch_next: Share::Helper {
+                seed: helper_sketch_next_seed,
+                length: 2 * input.len,
+            },
+        },
+    ])
+}
+
+/// The verification parameter used by the aggregators to evaluate the VDAF on a distributed input.
+pub struct VerifyParam {
+    /// Key used to derive the verification randomness from the nonce.
+    pub verify_rand_init: Key,
+}
+
+/// The VDAF setup algorithm that generates the verification parameters.
+#[cfg(test)]
+fn hits_setup(suite: Suite) -> Result<[VerifyParam; 2], VdafError> {
+    let verify_rand_init = Key::generate(suite)?;
+    Ok([
+        VerifyParam {
+            verify_rand_init: verify_rand_init.clone(),
+        },
+        VerifyParam {
+            verify_rand_init: verify_rand_init.clone(),
+        },
+    ])
+}
+
+/// The VDAF aggregation parameter, a sequence of equal-length candidate prefixes.
+#[derive(Debug, PartialEq)]
+pub struct AggregationParam<'a> {
+    prefixes: Vec<IdpfInput<'a>>,
+    len: usize,
+}
+
+impl<'a> AggregationParam<'a> {
+    /// Constructs an aggregation parameter from a sequence of prefixes. An error is returned if
+    /// the prefixes aren't equal length, a prefix appears twice in the set of prefixesx, or the
+    /// set of prefixes is empty.
+    pub fn new<P: IntoIterator<Item = IdpfInput<'a>>>(
+        prefixes: P,
+    ) -> Result<AggregationParam<'a>, VdafError> {
+        let mut unique_prefixes = HashSet::new();
+        let mut len = None;
+        let prefixes = prefixes
+            .into_iter()
+            .map(|prefix| {
+                if unique_prefixes.contains(&prefix) {
+                    return Err(VdafError::Uncategorized(format!(
+                        "prefix {:?} occurs twice in prefix set",
+                        prefix
+                    )));
+                }
+
+                if let Some(l) = len {
+                    if prefix.len != l {
+                        return Err(VdafError::Uncategorized(
+                            "prefixes must all have the same length".to_string(),
+                        ));
+                    }
+                } else {
+                    len = Some(prefix.len);
+                }
+                unique_prefixes.insert(prefix.clone());
```

Comment:
nit: does `unique_prefixes.insert(&prefix);` not work? You might need `P: IntoIterator(<Item = &IdpfInput<'a>>`, which may be a good idea anyway to avoid consuming `prefixes`.